File: c#/ProbabilityToFriendlyStringTests/FriendlyProbabilityTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using Xunit;

namespace Gregstoll.ProbabilityToFriendlyString
{
    [Serializable]
    public class FriendlyProbabilityTests
    {
        public struct TestCase
        {
            public int LineNumber { get; }
            public double Input { get; }
            public FriendlyProbability Output { get; }
            public TestCase(int lineNumber, double input, FriendlyProbability output)
            {
                LineNumber = lineNumber;
                Input = input;
                Output = output;
            }
        }

        public struct TestCaseFriendlyDescription
        {
            public int LineNumber { get; }
            public double Input { get; }
            public string ExpectedFriendlyDescription { get; }
            public TestCaseFriendlyDescription(int lineNumber, double input, string expectedFriendlyDescription)
            {
                LineNumber = lineNumber;
                Input = input;
                ExpectedFriendlyDescription = expectedFriendlyDescription;
            }
        }

        [Theory]
        [MemberData(nameof(GetTestCases))]
        public void TestsFromFile(TestCase testCase)
        {
            var actual = FriendlyProbability.FromProbability(testCase.Input);
            if (testCase.Output != actual)
            {
                Assert.True(false, string.Format("expected {0} actual {1} from line {2}", testCase.Output, actual, testCase.LineNumber));
            }
        }

        [Theory]
        [MemberData(nameof(GetFriendlyDescriptionTestCases))]
        public void FriendlyDescriptionTestsFromFile(TestCaseFriendlyDescription testCase)
        {
            var actual = FriendlyProbability.FromProbability(testCase.Input);
            if (testCase.ExpectedFriendlyDescription != actual.FriendlyDescription)
            {
                Assert.True(false, string.Format("expected {0} actual {1} from line {2}", testCase.ExpectedFriendlyDescription, actual.FriendlyDescription, testCase.LineNumber));
            }
        }

       public static IEnumerable<object[]> GetTestCases()
       {
            List<TestCase> testCases = new List<TestCase>();
            string path = Path.GetDirectoryName(typeof(FriendlyProbabilityTests).Assembly.Location);
            string testCasesPath = null;
            while (testCasesPath == null)
            {
                string tempPath = Path.Combine(path, "testCases.txt");
                if (File.Exists(tempPath))
                {
                    testCasesPath = tempPath;
                }
                path = Path.GetDirectoryName(path);
            }
            var lines = File.ReadAllLines(testCasesPath);
            int lineNumber = 0;
            foreach (string rawLine in lines)
            {
                ++lineNumber;
                string line = rawLine.Trim();
                if (line.StartsWith("#"))
                {
                    continue;
                }
                var parts = line.Split(',');
                double input = double.Parse(parts[0]);
                if (parts.Length == 3)
                {
                    testCases.Add(new TestCase(lineNumber, input, new FriendlyProbability(byte.Parse(parts[1]), byte.Parse(parts[2]), string.Empty)));
                }
                else if (parts.Length == 4)
                {
                    testCases.Add(new TestCase(lineNumber, input, new FriendlyProbability(byte.Parse(parts[1]), byte.Parse(parts[2]), string.Empty, parts[3])));
                }
                else
                {
                    throw new InvalidDataException(string.Format("line improperly formatted (line {0}: {1}", lineNumber, line));
                }
            }

            return testCases.Select(tc => new object[] { tc });
        }

       public static IEnumerable<object[]> GetFriendlyDescriptionTestCases()
       {
            List<TestCaseFriendlyDescription> testCases = new List<TestCaseFriendlyDescription>();
            string path = Path.GetDirectoryName(typeof(FriendlyProbabilityTests).Assembly.Location);
            string testCasesPath = null;
            while (testCasesPath == null)
            {
                string tempPath = Path.Combine(path, "testCases.friendlyDescription.txt");
                if (File.Exists(tempPath))
                {
                    testCasesPath = tempPath;
                }
                path = Path.GetDirectoryName(path);
            }
            var lines = File.ReadAllLines(testCasesPath);
            int lineNumber = 0;
            foreach (string rawLine in lines)
            {
                ++lineNumber;
                string line = rawLine.Trim();
                if (line.StartsWith("#"))
                {
                    continue;
                }
                var parts = line.Split(',');
                double input = double.Parse(parts[0]);
                testCases.Add(new TestCaseFriendlyDescription(lineNumber, input, parts[1]));
            }

            return testCases.Select(tc => new object[] { tc });
        }
    }
}

File: ruby/test/friendly_probability_tests.rb
require 'test/unit'
require_relative '../lib/probability_to_friendly_string'

class FriendlyProbabilityTests < Test::Unit::TestCase
    def test_all_cases
        directory = File.absolute_path(File.dirname(__FILE__))
        path = File.join(directory, 'testCases.txt')
        while !File.exists?(path)
            directory = File.dirname(directory)
            path = File.join(directory, 'testCases.txt')
        end
        lineNumber = 0
        File.open(path).each do |fullLine|
            lineNumber += 1
            line = fullLine.strip
            if (line.start_with?('#') or line.empty?)
                next
            end
            parts = line.split(',')
            case parts.length
            when 3
                expected = ProbabilityToFriendlyString::FriendlyProbability.new parts[1].to_i, parts[2].to_i, nil
            when 4
                expected = ProbabilityToFriendlyString::FriendlyProbability.new parts[1].to_i, parts[2].to_i, nil, parts[3]
            else
                flunk "Line badly formatted: #{line} (line #{lineNumber})"
            end
            actual = ProbabilityToFriendlyString::FriendlyProbability.fromProbability(parts[0].to_f)
            assert_equal expected.numerator, actual.numerator, "Numerator, called on #{parts[0]} (line #{lineNumber})"
            assert_equal expected.denominator, actual.denominator, "Denominator, called on #{parts[0]} (line #{lineNumber})"
            assert_equal expected.friendlyString, actual.friendlyString, "Denominator, called on #{parts[0]} (line #{lineNumber})"
        end
    end
end
